### PR TITLE
Prepend two ghosts

### PR DESCRIPTION
**Pull request by jokva**

### CODE REVIEW

File: src/tapeimage.cpp [hunk 1]
```diff
@@ -58,7 +58,7 @@ class address_map {
  * The record headers already read by tapeimage, stored in an order
  * (lower-address first fashion).
  *
- * A ghost node is inserted first:
+ * Two ghost node are inserted first:
```
Comment (achaikou) [inline src/tapeimage.cpp:61]: Two ghost node(s) are inserted

File: src/tapeimage.cpp [hunk 2]
```diff
@@ -255,8 +261,7 @@ record_index::find(std::int64_t n, iterator hint) const noexcept (false) {
         return hint;
     }
 
-    /* don't consider the first ghost node as a candidate */
-    const auto begin = std::next(this->begin());
+    const auto begin = this->begin();
     const auto end   = this->end();
```
Comment (achaikou) [inline src/tapeimage.cpp:264]: There actually seem to be not much need now for introduction of these variables

File: src/tapeimage.cpp [hunk 3]
```diff
@@ -645,7 +644,7 @@ void tapeimage::read_header_from_disk() noexcept (false) {
             this->recovery = LFP_PROTOCOL_TRYRECOVERY;
             head.prev = back2.next;
         }
-    } else if (this->recovery && !this->index.empty()) {
```
Comment (achaikou) [inline src/tapeimage.cpp:631]: You removed empty check.
Now imagine the test: I have a big file with two logical files and open file on the second logical file.
My base address is lets say 1000. But first header that is on position 1000 would be header.prev = 900.
If header type is not 0 or 1, recovery mode will be set, thus the check will fail. 

This check is valid only for the second header.

However saying this, you won't ever see this situation is real life. I actually believe that type =0 or =1 is the strongest check we have and in my opinion it shouldn't be soft fail with recovery, but hard fail. I do not believe think there is any recoverable file for which type is broken.

File: src/tapeimage.cpp [hunk 4]
```diff
@@ -525,14 +522,16 @@ std::int64_t tapeimage::readinto(void* dst, std::int64_t len) noexcept (false) {
...
     // TODO: consider when this says record, but physical file is EOF
     // TODO: end-of-file is an _empty_ record, i.e. two consecutive tape marks
     return this->current->type == tapeimage::file;
 }
 
 void tapeimage::read_header_from_disk() noexcept (false) {
-    assert(this->index.empty() or this->current == this->index.last());
+    try {
+        assert(this->index.last()->next == this->fp->tell());
+    } catch (const lfp::error&) {
```
Comment (achaikou) [inline src/tapeimage.cpp:533]: Maybe comment which error you are expecting here to ignore?
It's not very obvious.

File: src/tapeimage.cpp (hunk 2 above)
Comment (jokva) [reply src/tapeimage.cpp:264]: True, but it slightly cleans up the actual search, so I don't think it's so bad.

File: src/tapeimage.cpp (hunk 1 above)
Comment (jokva) [reply src/tapeimage.cpp:61]: Fix'd.

File: src/tapeimage.cpp (hunk 3 above)
Comment (jokva) [reply src/tapeimage.cpp:631]: Alright - I believe the solution to that is to re-introduce empty, but empty should not consider ghost nodes.

**Pull request merged**